### ISSUE TITLE
Compiling for ARM

### ISSUE THREAD

**Issue by SahanaSrinivas** [joshdoe/gst-plugins-vision#2]
I am using an  arm64 borad and i want to use gstreamer to access video from IDS ueye camera. I am very new to these topics. I have tried much but I have problems installing these plugins. Could you please guide me on the installation steps ?

**Comment by joshdoe** [comment]
You would need to compile the plugin for your architecture. I've only ever used these plugins on Windows/x86_64, but if the IDS uEye library supports ARM it should compile, perhaps with some minor fixes. Unfortunately I have no ARM experience or devices.

**Comment by harendracmaps** [comment]
While compiling on arm I am facig error for orc-0.4. what is the use of this library? Is it mandatory to be used.

**Comment by joshdoe** [comment]
ORC is for performance improvement, not strictly necessary, though used in
several other GStreamer libraries.

**Comment by harendracmaps** [comment]
Thanks for response. But it is failing while running cmake 
**cmake -G "Unix Makefiles" ..**
======
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
.

-- Checking for module 'orc-0.4'
--   No package 'orc-0.4' found
CMake Error at /usr/share/cmake-3.5/Modules/FindPkgConfig.cmake:367 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/share/cmake-3.5/Modules/FindPkgConfig.cmake:532 (_pkg_check_modules_internal)
  cmake/modules/FindOrc.cmake:16 (pkg_check_modules)
  CMakeLists.txt:49 (find_package)


-- Configuring incomplete, errors occurred!
See also "/persistent/gst-plugins-vision/build/CMakeFiles/CMakeOutput.log".

**Comment by 5shekel** [comment]
idsueyesrc does not use orc acceleration  
so for a "dirty" way. 
edit https://github.com/joshdoe/gst-plugins-vision/blob/ef66205b403c1e48c427035a813b39365886c796/CMakeLists.txt#L49
and remove the `REQUIRED` word.
maybe you would have to  skip (all?) other targets. 
remove lines 128-130 from CMakeLists.txt and replace with `add_subdirectory (sys/idsueye)`

i didnt test myslef as i dont have your hardware
but that should work,

**Comment by joshdoe** [comment]
Yes, Orc shouldn't be required, however it does seem to be supported on ARM, though the `FindOrc.cmake` file may need to be updated.

I do have an ARM platform that I haven't used for a long time, I might be able to try and get that setup again.

**Comment by joshdoe** [comment]
Just tried that, and looks like there are a few plugins that don't gracefully handle Orc being omitted. I'll try and make it optional. For now you can just uncomment the `add_subdirectory` lines to leave out extractcolor, misb, select, and aptina.

**Comment by harendracmaps** [comment]
> idsueyesrc does not use orc acceleration (yet)
> so for a "dirty" way.
> edit
> 
> https://github.com/joshdoe/gst-plugins-vision/blob/ef66205b403c1e48c427035a813b39365886c796/CMakeLists.txt#L49
> 
> 
> and remove the `REQUIRED` word.
> maybe you would have to skip (all?) other targets.
> remove lines 128-130 from CMakeLists.txt and replace with `add_subdirectory (sys/idsueye)`
> i didnt test myslef as i dont have your hardware
> but that should work, also there is probably a cleaner way to compile just one target...

Thanks!
I have already tried by removing REQUIRED and making TRUE -> FLASE in the next line.

EVEN I tried your suggestion "remove lines 128-130 from CMakeLists.txt and replace with add_subdirectory (sys/idsueye)"

Still faced the same issue
-----

-- Checking for module 'orc-0.4'
--   No package 'orc-0.4' found
CMake Error at /usr/share/cmake-3.5/Modules/FindPkgConfig.cmake:367 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/share/cmake-3.5/Modules/FindPkgConfig.cmake:532 (_pkg_check_modules_internal)
  cmake/modules/FindOrc.cmake:16 (pkg_check_modules)
  CMakeLists.txt:49 (find_package)


-- Configuring incomplete, errors occurred!
See also "/persistent/gst-plugins-vision/build/CMakeFiles/CMakeOutput.log".

**Comment by harendracmaps** [comment]
> Just tried that, and looks like there are a few plugins that don't gracefully handle Orc being omitted. I'll try and make it optional. For now you can just uncomment the `add_subdirectory` lines to leave out extractcolor, misb, select, and aptina.

Sorry but I could not follow you!

**Comment by 5shekel** [comment]
@harendracmaps try removing that orc line altogether.
strange t still complains even after you removed all other targets...

**Comment by 5shekel** [comment]
also try 
`sudo apt search liborc`

**Comment by harendracmaps** [comment]
> also try
> `sudo apt search liborc`

Thanks @5shekel ! liborc-0.4-dev library found for arm64. 
apt-get install liborc-0.4-dev resollved the orc issue but I am facing similar issue for BitFlow.

I want to use pylon module. Is it possible to comment lines from 52 till 110 and include only pylon?  I did that and faced issue as follow..
----------------
-- Could NOT find PYLON (missing:  PYLON_INCLUDE_DIR PYLON_LIBRARIES) 

-----------------------------------------------------------------------------
-- The following OPTIONAL packages could NOT be located on your system.
-- Consider installing them to enable more features from this software.
-----------------------------------------------------------------------------
   * Basler Pylon  <http://www.baslerweb.com/>
     Required to build Basler Pylon source element

-----------------------------------------------------------------------------

**Comment by joshdoe** [comment]
You probably need to edit the [FindPylon.cmake](https://github.com/joshdoe/gst-plugins-vision/blob/master/cmake/modules/FindPylon.cmake), I've only tested it on Windows. You've already installed the Pylon SDK? Find where `PylonC.h` is installed on your system, hopefully you just need to update `PYLON_DIR` and the `find_library` line, as it looks I have it hardcoded for Windows.

**Comment by joshdoe** [comment]
I've got Ubuntu 16.04 in a virtual machine now and am fixing things up. I've never cross-compiled, I might try that too.

**Comment by harendracmaps** [comment]
Thanks @joshdoe ! Waiting for the fixes.

**Comment by joshdoe** [comment]
@harendracmaps I'm pushing changes to the [linux](https://github.com/joshdoe/gst-plugins-vision/tree/linux) branch for now. Got it building the basic plugins, I'll try Pleora next.

**Comment by harendracmaps** [comment]
I have made changes for pylon..please let me know how to provide you patch? @joshdoe

**Comment by joshdoe** [comment]
Making changes now too, either do a pull request, or attach a patch/diff via [Gist](https://gist.github.com/) or similiar.

**Comment by harendracmaps** [comment]
https://gist.github.com/harendracmaps/a213d26a8da3ec2546a729cc62aa7b27

Here it is please review and merge it.  @joshdoe 

Thanks!

**Comment by joshdoe** [comment]
@harendracmaps I made almost the exact same changes, except I hardcoded `/opt/pylon5`, I don't seem to have `PYLON_ROOT` defined. Do you?

**Comment by harendracmaps** [comment]
yes while installation pylon I added it in my .bashrc.

Also I checked out linux branch and compiled. But it is failing while installing  
-------
Install the project...
-- Install configuration: ""
CMake Error at gst-libs/klv/cmake_install.cmake:36 (file):
  file INSTALL cannot find
  "/persistent/gst-plugins-vision/build/gst-libs/klv//libgstklv-1.0-0.pdb".
Call Stack (most recent call first):
  gst-libs/cmake_install.cmake:37 (include)
  cmake_install.cmake:37 (include)
 
Makefile:137: recipe for target 'install' failed
make: *** [install] Error 1

**Comment by joshdoe** [comment]
So... KLV, I should make that optional, as it's kind of experimental and only used with Pleora and the KLV plugins right now. Oh wait, that's the pdb file, @5shekel just put up some pull requests to disable PDB on Linux, as it's Microsoft specific. I'll add that next. For now, look [here](https://github.com/joshdoe/gst-plugins-vision/pull/18/commits/02e7419d8ed0ab61feb924133f25723ebbd20c83).

BTW, pushed changes to linux branch including the my version of your changes, except I didn't use PYLON_ROOT.

**Comment by harendracmaps** [comment]
Also it is creating libraries with prefix as "**liblib**" for example ..liblibgstpylon.so , liblibgstbayerutils.so, liblibgstklv-1.0-0.so

**Comment by joshdoe** [comment]
@harendracmaps yep, @5shekel pointed out the same thing. Fixed in linux branch.

**Comment by harendracmaps** [comment]
The libraries are getting installed in the path "/usr/include/gstreamer-1.0;/usr/lib/aarch64-linux-gnu/lib/gstreamer-1.0/"


Install the project...
-- Install configuration: ""
-- Installing: /usr/include/gstreamer-1.0;/usr/lib/aarch64-linux-gnu/gstreamer-1.0/include/../../lib/libgstklv-1.0-0.so
-- Installing: /usr/include/gstreamer-1.0;/usr/lib/aarch64-linux-gnu/gstreamer-1.0/include/../../lib/gstreamer-1.0/libgstbayerutils.so
-- Installing: /usr/include/gstreamer-1.0;/usr/lib/aarch64-linux-gnu/gstreamer-1.0/include/../../lib/gstreamer-1.0/libgstextractcolor.so
-- Installing: /usr/include/gstreamer-1.0;/usr/lib/aarch64-linux-gnu/gstreamer-1.0/include/../../lib/gstreamer-1.0/libgstklv.so
-- Set runtime path of "/usr/include/gstreamer-1.0;/usr/lib/aarch64-linux-gnu/gstreamer-1.0/include/../../lib/gstreamer-1.0/libgstklv.so" to ""
-- Installing: /usr/include/gstreamer-1.0;/usr/lib/aarch64-linux-gnu/gstreamer-1.0/include/../../lib/gstreamer-1.0/libgstmisb.so
-- Installing: /usr/include/gstreamer-1.0;/usr/lib/aarch64-linux-gnu/gstreamer-1.0/include/../../lib/gstreamer-1.0/libgstselect.so
-- Installing: /usr/include/gstreamer-1.0;/usr/lib/aarch64-linux-gnu/gstreamer-1.0/include/../../lib/gstreamer-1.0/libgstvideoadjust.so
-- Installing: /usr/include/gstreamer-1.0;/usr/lib/aarch64-linux-gnu/gstreamer-1.0/include/../../lib/gstreamer-1.0/libgstpylon.so
-- Set runtime path of "/usr/include/gstreamer-1.0;/usr/lib/aarch64-linux-gnu/gstreamer-1.0/include/../../lib/gstreamer-1.0/libgstpylon.so" to ""

# find /usr -name "libgstpylon.so"
/usr/include/gstreamer-1.0;/usr/lib/aarch64-linux-gnu/lib/gstreamer-1.0/libgstpylon.so

**Comment by joshdoe** [comment]
Yes, I've noticed that, a bad default. You can always change the `CMAKE_INSTALL_PREFIX`. See #20 also.

**Comment by 5shekel** [comment]
i install like this
```shell
mkdir Release && cd Release
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/opt/gst/plugins .. 
sudo make install 
#then let gstreamer know this is a plugin path
export GST_PLUGIN_PATH=/opt/gst/plugins
```
output (depends on your system found SDKs)
```
sudo make install
[  9%] Built target gstklv-1.0-0
[ 18%] Built target gstbayerutils
[ 31%] Built target gstextractcolor
[ 50%] Built target gstklv
[ 68%] Built target gstmisb
[ 77%] Built target gstselect
[ 90%] Built target gstvideoadjust
[100%] Built target gstidsueye
Install the project...
-- Install configuration: "Release"
-- Up-to-date: /opt/gst/plugins/lib/libgstklv-1.0-0.so
-- Up-to-date: /opt/gst/plugins/lib/gstreamer-1.0/libgstbayerutils.so
-- Up-to-date: /opt/gst/plugins/lib/gstreamer-1.0/libgstextractcolor.so
-- Up-to-date: /opt/gst/plugins/lib/gstreamer-1.0/libgstklv.so
-- Up-to-date: /opt/gst/plugins/lib/gstreamer-1.0/libgstmisb.so
-- Up-to-date: /opt/gst/plugins/lib/gstreamer-1.0/libgstselect.so
-- Up-to-date: /opt/gst/plugins/lib/gstreamer-1.0/libgstvideoadjust.so
-- Up-to-date: /opt/gst/plugins/lib/gstreamer-1.0/libgstidsueye.so
```

**Comment by joshdoe** [comment]
@5shekel and @harendracmaps check #20 for install issue, pushed commit to linux which hopefully makes everyone happy.

**Comment by ariel-elbit** [comment]
@joshdoe and @harendracmaps  I downloaded and installed basler pylon viewer 6.1.3 and I changed the path in the Findpylon.cmake file to match the current folders. I also changed set(_PREFIX "/usr/lib/aarch64-linux-gnu") (originally it was "/usr/lib"). after that I did make, and sudo make install:
[100%] Built target gstpylon
Install the project...
-- Install configuration: ""
-- Installing: /usr/lib/aarch64-linux-gnu/gstreamer-1.0/libgstbayerutils.so
-- Installing: /usr/lib/aarch64-linux-gnu/gstreamer-1.0/libgstextractcolor.so
-- Installing: /usr/lib/aarch64-linux-gnu/gstreamer-1.0/libgstmisb.so
-- Installing: /usr/lib/aarch64-linux-gnu/gstreamer-1.0/libgstselect.so
-- Installing: /usr/lib/aarch64-linux-gnu/gstreamer-1.0/libgstvideoadjust.so
-- Installing: /usr/lib/aarch64-linux-gnu/gstreamer-1.0/libgstpylon.so
-- Set runtime path of "/usr/lib/aarch64-linux-gnu/gstreamer-1.0/libgstpylon.so" to ""


But then when I check using gst-inspect-1.0 my pylon plugin doesn't appear and when I run:
"gst-launch-1.0 -v pylonsrc ! videoconvert ! video/x-raw,format=I420 ! x264enc tune=zerolatency bitrate=500 speed-preset=superfast ! rtph264pay ! udpsink host=127.0.0.1 port=5000"
I get: 
"WARNING: erroneous pipeline: no element "pylonsrc""

Please help me, I've been working on this for way too long and I can't manage to figure it out. I'm working on a jetson nano.
Thanks for your help. Ariel

**Comment by joshdoe** [comment]
@ariel-elbit , have you tried running gst-inspect on the absolute path of the plugin? I'm not familiar with ARM, but I'm guessing you set the prefix to where all the other plugins are located. Bump up the debugging levels with environment variables like GST_DEBUG=*:3

You might be missing a dependency, Pylon modules are already in your library path?

**Comment by ariel-elbit** [comment]
@joshdoe thanks for answering so quickly. What do you mean when you say to run the gst-inspect on the full path? Can you give an example?
Regarding the prefix, the answer is yes I did.
What does it mean to bump up the dubugging levels? Again, can you give an example?
And as far as I can tell the pylon modules are in the library path. There were some that were missing and I copied them into there. 
When I run gst-inspect-1.0 all of the pylon things are blacklisted.
Thanks again

**Comment by joshdoe** [comment]
@ariel-elbit run something like this to inspect a plugin directly, rather than relying on the standard plugin search path:
`gst-inspect-1.0 /usr/lib/aarch64-linux-gnu/gstreamer-1.0/libgstpylon.so`

Actually, that doesn't seem to work when I do that on Ubuntu, it works in Windows however.

You can increase the debug level when running GStreamer commands like this:
`GST_DEBUG=*:3 gst-inspect-1.0 pylon`

I usually use just Windows, but in a Ubuntu VM I can run this to see the shared library dependencies, and if they're found or not:
`ldd /usr/lib/aarch64-linux-gnu/gstreamer-1.0/libgstpylon.so`

Perhaps you're missing the Pylon dependencies in the library search path.

**Comment by ariel-elbit** [comment]
@joshdoe the ldd /usr/lib/aarch64-linux-gnu/gstreamer-1.0/libgstpylon.so showed me which file wasn't loading correctly, aparently I copied the wrong file into the usr/lib/aarch64-linux-gnu/gstreamer-1.0 folder. After building the gst-plugins project I needed to copy the libgstpylon.so file into the usr/lib/aarch64-linux-gnu/gstreamer-1.0 folder. After that it it recognized the plugin. (maybe you can add that part to the installation instructions)
After that, when I run:
gst-launch-1.0 -v pylonsrc ! videoconvert ! video/x-raw,format=I420 ! videoconvert ! ximagesink, the streaming stops a couple of seconds after it starts:
Setting pipeline to PAUSED ...
Pipeline is live and does not need PREROLL ...
Setting pipeline to PLAYING ...
New clock: GstSystemClock
/GstPipeline:pipeline0/GstPylonSrc:pylonsrc0.GstPad:src: caps = video/x-raw, format=(string)RGB, width=(int)1920, height=(int)1080, framerate=(fraction)30/1, pixel-aspect-ratio=(fraction)1/1
/GstPipeline:pipeline0/GstVideoConvert:videoconvert0.GstPad:src: caps = video/x-raw, width=(int)1920, height=(int)1080, framerate=(fraction)30/1, pixel-aspect-ratio=(fraction)1/1, format=(string)I420
/GstPipeline:pipeline0/GstCapsFilter:capsfilter0.GstPad:src: caps = video/x-raw, width=(int)1920, height=(int)1080, framerate=(fraction)30/1, pixel-aspect-ratio=(fraction)1/1, format=(string)I420
/GstPipeline:pipeline0/GstVideoConvert:videoconvert1.GstPad:src: caps = video/x-raw, width=(int)1920, height=(int)1080, framerate=(fraction)30/1, pixel-aspect-ratio=(fraction)1/1, format=(string)BGRx
/GstPipeline:pipeline0/GstXImageSink:ximagesink0.GstPad:sink: caps = video/x-raw, width=(int)1920, height=(int)1080, framerate=(fraction)30/1, pixel-aspect-ratio=(fraction)1/1, format=(string)BGRx
/GstPipeline:pipeline0/GstVideoConvert:videoconvert1.GstPad:sink: caps = video/x-raw, width=(int)1920, height=(int)1080, framerate=(fraction)30/1, pixel-aspect-ratio=(fraction)1/1, format=(string)I420
/GstPipeline:pipeline0/GstCapsFilter:capsfilter0.GstPad:sink: caps = video/x-raw, width=(int)1920, height=(int)1080, framerate=(fraction)30/1, pixel-aspect-ratio=(fraction)1/1, format=(string)I420
/GstPipeline:pipeline0/GstVideoConvert:videoconvert0.GstPad:sink: caps = video/x-raw, format=(string)RGB, width=(int)1920, height=(int)1080, framerate=(fraction)30/1, pixel-aspect-ratio=(fraction)1/1
/GstPipeline:pipeline0/GstPylonSrc:pylonsrc0.GstPad:src: caps = video/x-raw, format=(string)RGB, width=(int)1855, height=(int)1056, framerate=(fraction)0/1, pixel-aspect-ratio=(fraction)1/1
/GstPipeline:pipeline0/GstVideoConvert:videoconvert0.GstPad:src: caps = video/x-raw, width=(int)1855, height=(int)1056, framerate=(fraction)0/1, pixel-aspect-ratio=(fraction)1/1, format=(string)I420
/GstPipeline:pipeline0/GstCapsFilter:capsfilter0.GstPad:src: caps = video/x-raw, width=(int)1855, height=(int)1056, framerate=(fraction)0/1, pixel-aspect-ratio=(fraction)1/1, format=(string)I420
/GstPipeline:pipeline0/GstVideoConvert:videoconvert1.GstPad:src: caps = video/x-raw, width=(int)1855, height=(int)1056, framerate=(fraction)0/1, pixel-aspect-ratio=(fraction)1/1, format=(string)BGRx
/GstPipeline:pipeline0/GstXImageSink:ximagesink0.GstPad:sink: caps = video/x-raw, width=(int)1855, height=(int)1056, framerate=(fraction)0/1, pixel-aspect-ratio=(fraction)1/1, format=(string)BGRx
/GstPipeline:pipeline0/GstVideoConvert:videoconvert1.GstPad:sink: caps = video/x-raw, width=(int)1855, height=(int)1056, framerate=(fraction)0/1, pixel-aspect-ratio=(fraction)1/1, format=(string)I420
/GstPipeline:pipeline0/GstCapsFilter:capsfilter0.GstPad:sink: caps = video/x-raw, width=(int)1855, height=(int)1056, framerate=(fraction)0/1, pixel-aspect-ratio=(fraction)1/1, format=(string)I420
/GstPipeline:pipeline0/GstVideoConvert:videoconvert0.GstPad:sink: caps = video/x-raw, format=(string)RGB, width=(int)1855, height=(int)1056, framerate=(fraction)0/1, pixel-aspect-ratio=(fraction)1/1
ERROR: from element /GstPipeline:pipeline0/GstPylonSrc:pylonsrc0: Internal data stream error.
Additional debug info:
gstbasesrc.c(3055): gst_base_src_loop (): /GstPipeline:pipeline0/GstPylonSrc:pylonsrc0:
streaming stopped, reason error (-5)
Execution ended after 0:00:00.620270229
Setting pipeline to PAUSED ...
Setting pipeline to READY ...
Setting pipeline to NULL ...
Freeing pipeline ...

Any ideas to why?

**Comment by joshdoe** [comment]
Try to increase the debugging level, like this:
```GST_DEBUG=*:3,pylon*:6 gst-launch-1.0 -v pylonsrc ! videoconvert ! ximagesink```

That will have a lot of output, you might need to attach it as a file instead of pasting it in the message.

BTW, you might have a reason for doing so, but the pipeline you showed can be simplified to what I just wrote above, only one videoconvert is needed and no need to specify caps. If your camera has multiple color formats and you're trying to specifically pull out the I420 format, then you'd want to use a capsfilter instead.

**Comment by pdeman** [comment]
Hi, I need to compile idsueyesrc on a tx2.
Have you found a way to do it ?
is there a possibility to compile only idsueyesrc and not others ?

**Comment by pdeman** [comment]
I have finally made it works but I have a lot of difficulty to connect it to others pluging like nvvconv , queue etc ... problems that I don't have on windows.
